File: 881-boats-to-save-people/881-boats-to-save-people.cpp
class Solution {
public:

    int numRescueBoats(vector<int>& people, int limit) {
        
        vector<int> mp(30001, 0);
        for(auto it:people) {
            mp[it]++;
        }
        
        int count=0;
        for(int i=0;i<people.size();i++) {
            if(mp[people[i]]>0) {
                int target = limit-people[i];
                mp[people[i]]--;
                while(mp[target]<=0 && target>=1) {
                    target--;
                }
                count++;
                if(target>=1) {
                    mp[target]--;
                }  
            }
        }
        return count   ;
    }
};

File: 881-boats-to-save-people/881-boats-to-save-people.java
class Solution {
    public int numRescueBoats(int[] people, int limit) {
        int[] map = new int[limit + 1];
        
        for(int x : people){
            map[x]++;
        }
        
        int i = 0;
        int j = limit;
        int numPeople = 0;
        
        while(i <= j){
            while(map[i] == 0 && i <= j){
                i++;
            }
            while(map[j] == 0 && i <= j){
                j--;
            }
            if(i + j <= limit){
                if(map[j] > 0)
                   map[j]--;
                if(map[i] > 0)
                   map[i]--;
            }
            else{
                if(map[j] > 0)
                   map[j]--;
            }
            numPeople++;
        }
        return numPeople - 1;
    }
}